swaybar/tray/dbus.c

```diff
@@ -0,0 +1,131 @@
+#define _XOPEN_SOURCE 500
+
+#ifdef __FreeBSD__
+#include <dev/evdev/input-event-codes.h>
+#else
+#include <linux/input-event-codes.h>
+#endif
+#include <poll.h>
+#include <time.h>
+#include <wlr/util/log.h>
+#include "swaybar/event_loop.h"
+#include "swaybar/tray/dbus.h"
+#include "swaybar/tray/sni_host.h"
+#include "swaybar/tray/sni_watcher.h"
+
+static const char *fd_iface = "org.freedesktop.StatusNotifierWatcher";
+static const char *fd_notifier_host = "org.freedesktop.StatusNotifierHost";
+
+bool dbus_init() {
+	int ret = 0;
+	sd_bus *bus = NULL;
+	/* BUG: *slot should contained in sni_watcher struct to facilitate multiple
+	 * watchers but when done in this way the sd-bus handler callbacks 
+	 * (handle_new_icon) were getting bad userdata
+	 */
+
+	sd_bus_slot *slot = NULL;
+	// TODO: Add other watchers for KDE, etc.
+	struct sni_watcher fd_sni_watcher = {0}; 
```

Comment:
Your problem is (probably) not about the slot being a field of the watcher struct, it's that you've declared the watcher on the stack, meaning that when the function returns, it goes out of scope, causing its memory to become corrupted when it is used in the callback later.

The solution is to `calloc` it on the heap.